engine/src/main/java/org/terasology/particles/ParticleSystemStateData.java

```diff
@@ -0,0 +1,144 @@
+ * limitations under the License.
+ */
+package org.terasology.particles;
+
+import com.google.common.collect.BiMap;
+import org.terasology.entitySystem.Component;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.logic.location.LocationComponent;
+import org.terasology.particles.components.ParticleEmitterComponent;
+import org.terasology.particles.components.ParticleSystemComponent;
+import org.terasology.particles.functions.affectors.AffectorFunction;
+import org.terasology.particles.functions.generators.GeneratorFunction;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Holds the data of a particle system used to render the system.
+ * A particle system's state data is updated whenever the particle system is changed.
+ */
+public class ParticleSystemStateData {
+
+    public final EntityRef entityRef;
+    public final ParticlePool particlePool;
+    public int collisionUpdateIteration;
+
+    // Data used by the updater implementation, fetched every update, changes when the system is changed.
+    public final FetchedData fetchedData;
+
+    public ParticleSystemStateData(final EntityRef entityRef) {
+        this.entityRef = entityRef;
+        this.particlePool = new ParticlePool(entityRef.getComponent(ParticleSystemComponent.class).particlePoolSize);
+        this.collisionUpdateIteration = 0;
+
+        this.fetchedData = new FetchedData();
+    }
+
+    public class FetchedData {
+
+        // Affector functions called on each particle during update
+        public final Map<Component, AffectorFunction> affectors;
+
+        public ParticleSystemComponent particleSystemComponent;
+        public List<EmitterData> emitters;
+
+        public FetchedData() {
+            this.affectors = new LinkedHashMap<>();
+            this.particleSystemComponent = entityRef.getComponent(ParticleSystemComponent.class);
+        }
+
+        /*
+        * Creates EmitterData for each emitter to be used by updater, called when emitter changes (add/remove/generator change)
+        * */
+        public void update(final BiMap<Class<Component>, GeneratorFunction> generatorFunctions,
+                           final BiMap<Class<Component>, AffectorFunction> affectorFunctions
+        ) {
+
+            if (emitters == null) {
+                emitters = new ArrayList<>(particleSystemComponent.getEmitters().size());
+            } else {
+                emitters.clear();
+            }
+
+            // Thanks IntelliJ ;)
+            emitters.addAll(particleSystemComponent.getEmitters().stream().map(emitter -> new EmitterData(
+                    emitter != null ? emitter.getComponent(ParticleEmitterComponent.class) : null,
+                    emitter != null ? emitter.getComponent(LocationComponent.class) : null,
+                    generatorFunctions
+            )).collect(Collectors.toList()));
```

Comment:
I'm not sure about the following snippet (maybe I throw out too much when the emitter is `null`). Having worked with Scala for some time this line is quite logical to me, but I get why it might be confusing to others. 

```java
emitters.addAll(
  particleSystemComponent.getEmitters().stream()
    .filter(Objects::nonNull)
    .map(emitter -> new EmitterData(emitter.getComponent(...), emitter.getComponent(...), generatorFncs))
    .collect(Collectors.toList()));
```